### PR TITLE
Wide&Deep from TensorFlow to PyTorch

### PR DESCRIPTION
**Pull request by miguelgfierro**
### Description

TF: {'rmse': 0.9526174435076816, 'mae': 0.7576649205052852} and {'ndcg_at_k': 0.08594396229187452, 'precision_at_k': 0.08557794273594911}
PT: {'rmse': 0.9477648484822748, 'mae': 0.7491909999155998} and {'ndcg_at_k': 0.09333924630101367, 'precision_at_k': 0.09023354564755841}

### Related Issues

Fix https://github.com/recommenders-team/recommenders/issues/2072
Related to #2073

### Checklist:

- [ ] I have followed the [contribution guidelines](https://github.com/recommenders-team/recommenders/blob/main/CONTRIBUTING.md) and code style for this project.
- [ ] I have added tests covering my contributions.
- [ ] I have updated the documentation accordingly.
- [ ] I have [signed the commits](https://github.com/recommenders-team/recommenders/wiki/How-to-sign-commits), e.g. `git commit -s -m "your commit message"`.
- [ ] This PR is being made to `staging branch` AND NOT TO `main branch`.

### CODE REVIEW

**Comment by miguelgfierro** [comment]
continuation of this PR https://github.com/recommenders-team/recommenders/pull/2168

File: recommenders/models/wide_deep/wide_deep_utils.py [hunk 1]
```diff
@@ -1,10 +1,167 @@
...
+                    layers.append(nn.Dropout(dnn_dropout))
+                in_dim = units
+            layers.append(nn.Linear(in_dim, 1))
+            self.dnn = nn.Sequential(*layers)
+
+    @staticmethod
+    def _cross_hash(user_ids, item_ids, bucket_size):
+        """Deterministic hash of (user, item) pairs into ``[0, bucket_size)``."""
+        return (
+            (user_ids.long() * 6364136223846793005 + item_ids.long()) % bucket_size
```
Comment (anargyri) [inline recommenders/models/wide_deep/wide_deep_utils.py]: 6364136223846793005 is kind of a "magic number", is there a better way to define it? (also better to define it as a parameter at the beginning)

Comment (miguelgfierro) [reply recommenders/models/wide_deep/wide_deep_utils.py]: fixed!

File: recommenders/models/wide_deep/wide_deep_utils.py [hunk 2]
```diff
@@ -1,10 +1,169 @@
...
+    DEFAULT_RATING_COL,
+    DEFAULT_PREDICTION_COL,
+)
+
+logger = logging.getLogger(__name__)
+
+# Multiplier from Knuth's 64-bit linear congruential generator (LCG),
+# used to hash (user, item) pairs into crossed-feature buckets.
+# Source: Knuth, "The Art of Computer Programming", Vol. 2, 1997.
+CROSS_HASH_MULTIPLIER = 6364136223846793005
```
Comment (anargyri) [inline recommenders/models/wide_deep/wide_deep_utils.py]: This is still a magic number and looks like just an arbitrary large number. It should probably be something related to `n_items`. Alternatively, you could hash the (user, item) tuple like [this](https://docs.python.org/3.14/reference/datamodel.html#object.__hash__).

File: recommenders/models/wide_deep/wide_deep_utils.py [hunk 3]
```diff
@@ -19,195 +178,401 @@ def build_feature_columns(
...
+    elif has_wide:
+        model_type = "wide"
+    else:
+        model_type = "deep"
+
+    # Merge user/item counts from whichever config is present
+    cfg = wide_columns if has_wide else deep_columns
+    n_users = cfg["n_users"]
+    n_items = cfg["n_items"]
+    crossed_feat_dim = wide_columns.get("crossed_feat_dim", 1000)
```
Comment (anargyri) [inline recommenders/models/wide_deep/wide_deep_utils.py]: This 1000 should be 0, right? (it corresponds to the case there are no wide columns)

File: recommenders/models/wide_deep/wide_deep_utils.py [hunk 4]
```diff
@@ -19,195 +178,401 @@ def build_feature_columns(
...
+    user_dim = deep_columns.get("user_dim", 8)
```
Comment (anargyri) [inline recommenders/models/wide_deep/wide_deep_utils.py]: Also 8 is 0 if user / item dim is not applicable.

File: recommenders/models/wide_deep/wide_deep_utils.py [hunk 5]
```diff
@@ -19,195 +178,401 @@ def build_feature_columns(
...
+    loss_fn = nn.MSELoss()
+    model.train()
+
+    step = 0
+    while step < steps:
+        # Create a fresh iterator each epoch so data is reshuffled
+        for batch in loader:
+            step += 1
+            if step > steps:
+                break
```
Comment (anargyri) [inline recommenders/models/wide_deep/wide_deep_utils.py]: This conditional is redundant because of the while statement.

File: tests/unit/recommenders/models/test_wide_deep_utils.py [hunk 6]
```diff
@@ -49,92 +51,149 @@ def pd_df():
...
-    model.train(
-        input_fn=pandas_input_fn(
-            df=data,
-            y_col=DEFAULT_RATING_COL,
-            batch_size=1,
-            num_epochs=None,
-            shuffle=True,
-        ),
+    # Wide config should be populated, deep should be empty
+    assert wide_columns
```
Comment (anargyri) [inline tests/unit/recommenders/models/test_wide_deep_utils.py]: We could keep the assert that checks for the length of wide_columns.

**Comment by miguelgfierro** [comment]
I'm remaking the model to have a better interface. I tried to compare apples to apples and with 21 epochs, which is around the same 50000 steps we had before, the metric numbers are the same:

wide_deep_movielens.ipynb — updated to use new API, metrics match closely:
  - RMSE: 0.9478 → 0.9488
  - MAE: 0.7492 → 0.7523
  - NDCG@10: 0.0933 → 0.0930
  - Precision@10: 0.0902 → 0.0902

**Comment by miguelgfierro** [comment]
@anargyri I have changed the interfaces that were a little bit off and simplified everything. Please take a look

**Review by anargyri — APPROVED** [review]

**Comment by miguelgfierro** [comment]
The errors in the tests are not related to this PR but an issue with Wikidata API that is being fixed at https://github.com/recommenders-team/recommenders/pull/2284
